include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,516 @@
+
+        return it->second;
+    }
+
+    /**
+     * Iterates over the real PIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function.
+     * @param fn The function to be applied to each primary input.
+     */
+    template <typename Fn>
+    void foreach_real_pi(Fn&& fn)
+    {
+        static_cast<Ntk*>(this)->foreach_pi(
+            [&](const auto& i)
+            {
+                if (!is_virtual_pi(i))
+                {
+                    std::forward<Fn>(fn)(i);
+                }
+            });
+    }
+
+    /**
+     * Iterates over the virtual PIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function.
+     * @param fn The function to be applied to each primary input.
+     */
+    template <typename Fn>
+    void foreach_virtual_pi(Fn&& fn) const
+    {
+        mockturtle::detail::foreach_element(_storage->virtual_inputs.cbegin(), _storage->virtual_inputs.cend(), fn);
+    }
+
+    /**
+     * Iterates over the virtual CIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function to be applied.
+     * @param fn The function to be applied.
+     */
+    template <typename Fn>
+    void foreach_real_ci(Fn&& fn)
+    {
+        static_cast<Ntk*>(this)->foreach_ci(
+            [&](const auto& i)
+            {
+                if (!is_virtual_ci(i))
+                {
+                    std::forward<Fn>(fn)(i);
+                }
+            });
+    }
+
+    /**
+     * Iterates over the virtual CIs of the circuit and applies a given function.
+     *
+     * @tparam Fn The type of the function.
+     * @param fn The function to be applied to each primary input.
+     */
+    template <typename Fn>
+    void foreach_virtual_ci(Fn&& fn) const
+    {
+        mockturtle::detail::foreach_element(_storage->virtual_inputs.cbegin(), _storage->virtual_inputs.cend(), fn);
+    }
+
+    /**
+     * Removes virtual input nodes from the virtual_pi_network.
+     *
+     * This function removes the virtual input nodes from the network by substituting them with their corresponding
+     * real input nodes. It then performs a cleanup to remove any dangling PIs..
+     * Finally, it clears the virtual_inputs and map_virt_to_real_pi data structures in the _storage object.
+     */
+    void remove_virtual_input_nodes()
+    {
+        for (const auto& map_item : _storage->map_virt_to_real_pi)
+        {
+            Ntk::substitute_node(Ntk::get_node(map_item.first), map_item.second);
+        }
+
+        *this = mockturtle::cleanup_dangling(*this, 1);
+
+        // Clear virtual_inputs after using it
+        _storage->virtual_inputs.clear();
+        _storage->map_virt_to_real_pi.clear();
+    }
+
+  protected:
+    /**
+     * Shared pointer if the virtual PI storage.
+     */
+    std::shared_ptr<virtual_storage<Ntk>> _storage;
```

Comment:
```suggestion
    storage strg;
```
if you add ``using storage = std::shared_ptr<virtual_storage<Ntk>>;`` in the first line after ``public``. This is how we do it in _fiction_.